### ISSUE TITLE
[Feature Request] Queries

### ISSUE THREAD

**Issue by Elvecent** [temporalio/sdk-core#476]
### Is your feature request related to a problem? Please describe.

I'm trying to use this library in Rust and implement Workflows that have some queryable state. However, I can't find a way to make or handle queries, it seems there's something lacking here. Either that or I'm missing something. All I could find was [respond_legacy_query](https://github.com/temporalio/sdk-core/blob/master/core/src/worker/client.rs#L107), what's that by the way?

### Describe the solution you'd like

An API akin to what's there in Go code samples: 
https://legacy-documentation-sdks.temporal.io/go/how-to-use-queries-in-go
https://legacy-documentation-sdks.temporal.io/go/how-to-handle-a-query-in-a-workflow-in-go

**Comment by Sushisource** [comment]
@Elvecent Indeed, they're not implemented in the Rust SDK itself.

The Rust SDK in this repo exists only for internal testing purposes at this point and comes with no API stability guarantees. As such I probably won't be able to prioritize adding arbitrary query support to it unless I happen to have a reason to need it for testing.

Once we've prioritized releasing a production ready Rust SDK, I plan on changing the API quite a bit, to be trait based, and the main reason I haven't added any real query support yet is because it looks quite funky without a very different API.

If you wanted to contribute something which looks like:
```rust
impl WfContext {
    fn register_query(&self, query_name: impl Into<String>, query_fn: impl Fn() -> Payload) {}
}
```

I'd be open to that as an intermediate solution

**Comment by Elvecent** [comment]
Thanks for the explanation. It's presently not a pressing need for me, but if that changes, I'll look into it.

**Comment by remeq** [comment]
@Sushisource is an official Temporal Rust SDK library on a roadmap for future ?

**Comment by Sushisource** [comment]
@remeq See the blurb at the end of the README [here](https://github.com/temporalio/sdk-core#the-rust-sdk). We do not have official plans to productionize yet.

**Comment by awkure** [comment]
I was going to implement them myself and have a couple of questions regarding that. 
1. How much and what exactly is left for them to be implemented? I've seen there were some developments in https://github.com/temporalio/sdk-core/pull/135 but it appears that they're still unimplemented in [WorkflowFuture::handle_job](https://github.com/temporalio/sdk-core/blob/6f447a6edb0387db196b939de966c02a729ff269/sdk/src/workflow_future.rs#L184-L189) endpoint.
2. I tried doing it with the analogy with signal jobs, but couldn't find where [these signals](https://github.com/temporalio/sdk-core/blob/6f447a6edb0387db196b939de966c02a729ff269/sdk/src/workflow_future.rs#L112) (besides tests) are consumed. It'll be really helpful knowing that.

**Comment by awkure** [comment]
Ok, the second part was clarified. The first question is still open.

**Comment by Elvecent** [comment]
It would seem that at least we would need to add
1. A hashmap like `sig_chans` to  `WorkflowFuture` that would map query names to `impl Fn() -> Payload` callbacks
2. A `QueryWorkflow` command handler in `WorkflowFuture::handle_job` that looks up in said hashmap and executes the callback
3. A `SubscribeQuery` variant to `RustWfCmd` enum together with a corresponding match branch
4. A `register_workflow` method for `WfContext` that populates the hashmap by calling `RustWfCmd::SubscribeQuery`

If I'm not missing something non-trivial, this should be rather straightforward.

**Comment by Sushisource** [comment]
@Elvecent Yep that's pretty much all right (though I think in 4 you meant to name it `register_query_handler` or similar).

**Comment by Elvecent** [comment]
I think a hit the gotcha. I don't see a way to complete the query request after the result is computed in `WorkflowFuture::handle_job`

**Comment by Sushisource** [comment]
@Elvecent I see, yes you'll need to plumb some way to include query responses in the outgoing commands for completion.

Probably the easiest thing to do would be to simply store a new list of query responses on the `WorkflowFuture` itself, and then drain those and include them in the completion inside of `send_completion`

**Comment by Elvecent** [comment]
@Sushisource Could you please clarify what do you refer to as `send_completion`? I'm not sure where to look for it

**Comment by Sushisource** [comment]
> @Sushisource Could you please clarify what do you refer to as `send_completion`? I'm not sure where to look for it

It's a method defined on `WorkflowFuture`

**Comment by c-thiel** [comment]
@Elvecent  did you ever manage to finish your implementation or have a work in progress somewhere that you could share?

**Comment by Elvecent** [comment]
@c-thiel here's what I have: https://github.com/Elvecent/sdk-core/tree/queries
I ended up not needing it, though, so it's hardly tested
